CRM/PivotReport/AbstractData.php

```diff
@@ -0,0 +1,526 @@
+   *
+   * @return int
+   */
+  public function rebuildData($cacheGroup, $entityName, array $params, $offset = 0, $multiValuesOffset = 0, $page = 0) {
+    $total = $this->getCount($params);
+    $apiParams = $this->getEntityApiParams($params);
+    $index = NULL;
+    $count = 0;
+
+    while ($offset < $total) {
+      if ($offset) {
+        $offset--;
+      }
+
+      $pages = $this->getPaginatedResults($entityName, $apiParams, $offset, $multiValuesOffset, $page, $index);
+
+      $count += $this->cachePages($cacheGroup, $pages);
+
+      $lastPageIndex = count($pages) - 1;
+      $offset = $pages[$lastPageIndex]->getNextOffset();
+      $multiValuesOffset = $pages[$lastPageIndex]->getNextMultiValuesOffset();
+      $page = $pages[$lastPageIndex]->getPage() + 1;
+      $index = $pages[$lastPageIndex]->getIndex();
+    }
+
+    return $count;
+  }
+
+  /**
+   * Rebuilds entity header cache.
+   *
+   * @param \CRM_PivotCache_AbstractGroup $cacheGroup
+   * @param array $header
+   */
+  public function rebuildHeader($cacheGroup, array $header) {
+    $cacheGroup->cacheHeader($header);
+  }
+
+  /**
+   * Returns an array of entity data pages.
+   *
+   * @param string $entityName
+   * @param array $apiParams
+   * @param int $offset
+   * @param int $multiValuesOffset
+   * @param int $page
+   * @param string $index
+   *
+   * @return int
+   */
+  protected function getPaginatedResults($entityName, array $apiParams, $offset = 0, $multiValuesOffset = 0, $page = 0, $index = NULL) {
+    $result = array();
+    $rowsCount = 0;
+
+    $apiParams['options']['offset'] = $offset;
+
+    $entities = civicrm_api3($entityName, 'get', $apiParams);
+
+    $formattedEntities = $this->formatResult($entities['values']);
+
+    unset($entities);
+
+    while (!empty($formattedEntities)) {
+      $split = $this->splitMultiValues($formattedEntities, $offset, $multiValuesOffset);
+      $rowsCount += count($split['data']);
+
+      if ($rowsCount > self::ROWS_PAGINATED_LIMIT) {
+        break;
+      }
+
+      if ($split['info']['index'] !== $index) {
+        $page = 0;
+        $index = $split['info']['index'];
+      }
+
+      $result[] = new CRM_PivotReport_DataPage($split['data'], $index, $page++, $split['info']['nextOffset'], $split['info']['multiValuesOffset']);
+
+      unset($split['data']);
+
+      $formattedEntities = array_slice($formattedEntities, $split['info']['nextOffset'] - $offset, NULL, TRUE);
+
+      $offset = $split['info']['nextOffset'];
+      $multiValuesOffset =  $split['info']['multiValuesOffset'];
+    }
+
+    return $result;
+  }
+
+  /**
+   * Puts an array of pages into cache.
+   *
+   * @param \CRM_PivotCache_AbstractGroup $cacheGroup
+   * @param array $pages
+   *
+   * @return int
+   */
+  protected function cachePages($cacheGroup, array $pages) {
+    $count = 0;
+
+    foreach ($pages as $page) {
+      $count += $cacheGroup->cachePacket($page->getData(), $page->getIndex(), $page->getPage());
```

Comment:
I wonder if it wouldn't make more sense for the `cachePacket` method to be renamed to `cachePage` and then you would just pass the page object to it